src/daft-io/src/local.rs

```diff
@@ -84,49 +107,109 @@ pub struct LocalFile {
+        const LOCAL_PROTOCOL: &str = "file://";
+        if let Some(uri) = uri.strip_prefix(LOCAL_PROTOCOL) {
             Ok(GetResult::File(LocalFile {
-                path: path.to_path_buf(),
+                path: uri.into(),
                 range,
             }))
         } else {
-            return Err(Error::InvalidFilePath {
-                path: uri.to_string(),
-            }
-            .into());
+            Err(Error::InvalidFilePath { path: uri.into() }.into())
         }
     }
 
     async fn get_size(&self, uri: &str) -> super::Result<usize> {
-        const TO_STRIP: &str = "file://";
-        if let Some(p) = uri.strip_prefix(TO_STRIP) {
-            let path = std::path::Path::new(p);
-            let file = tokio::fs::File::open(path)
-                .await
-                .context(UnableToOpenFileSnafu {
-                    path: path.to_string_lossy(),
-                })?;
-            let metadata = file.metadata().await.context(UnableToOpenFileSnafu {
-                path: path.to_string_lossy(),
-            })?;
-            return Ok(metadata.len() as usize);
-        } else {
-            return Err(Error::InvalidFilePath {
+        const LOCAL_PROTOCOL: &str = "file://";
+        let Some(uri) = uri.strip_prefix(LOCAL_PROTOCOL) else {
+            return Err(Error::InvalidFilePath { path: uri.into() }.into());
+        };
+        let meta = tokio::fs::metadata(uri)
+            .await
+            .context(UnableToFetchFileMetadataSnafu {
                 path: uri.to_string(),
-            }
-            .into());
-        }
+            })?;
+        Ok(meta.len() as usize)
     }
 
     async fn ls(
         &self,
-        _path: &str,
+        path: &str,
         _delimiter: Option<&str>,
         _continuation_token: Option<&str>,
     ) -> super::Result<LSResult> {
-        unimplemented!("local ls");
+        let mut s = self.iter_dir(path, None, None).await?;
+        let size = s.size_hint().1.unwrap_or(0);
+        let mut files = Vec::with_capacity(size);
+        while let Some(file_meta) = StreamExt::next(&mut s).await.transpose()? {
+            files.push(file_meta);
+        }
+        Ok(LSResult {
+            files,
+            continuation_token: None,
+        })
+    }
+
+    async fn iter_dir(
+        &self,
+        uri: &str,
+        _delimiter: Option<&str>,
+        _limit: Option<usize>,
+    ) -> super::Result<BoxStream<super::Result<FileMetadata>>> {
+        const LOCAL_PROTOCOL: &str = "file://";
+        let Some(uri) = uri.strip_prefix(LOCAL_PROTOCOL) else {
+            return Err(Error::InvalidFilePath { path: uri.into() }.into());
+        };
+        let meta =
+            tokio::fs::metadata(uri)
+                .await
+                .with_context(|_| UnableToFetchFileMetadataSnafu {
+                    path: uri.to_string(),
+                })?;
+        let uri = uri.to_string();
+        let s = stream! {
```

Comment:
I would actually break this up to 2 streams that you can return rather than putting defining a custom `stream!`
```
if meta.file_type().is_file() {
  futures:stream:iter([my_single_file])
} else {
  let dir_entries = ...
  let mut dir_stream = ...
  dir_stream.then(...).boxed()
}
```